crates/symbolicator/src/services/symbolication.rs

```diff
@@ -130,76 +130,99 @@ impl<'a> SymCacheLookupResult<'a> {
-/// modules determined to be referenced by this minidump by
-/// [`SymbolicationActor::get_referenced_modules_from_minidump`].  It contains the CFI cache
+/// modules determined to be referenced by this minidump.  It contains the CFI cache
 /// status of the modules and allows loading the CFI from the caches for the correct
 /// minidump stackwalking.
 ///
 /// It maintains the status of the object file availability itself as well as any features
 /// provided by it.  This can later be used to compile the required modules information
 /// needed for the final response on the JSON API.  See the [`ModuleListBuilder`] struct for
 /// this.
-#[derive(Debug, Serialize, Deserialize)]
+#[derive(Clone, Debug)]
 struct CfiCacheModules {
+    cache_files: Vec<Arc<CfiCacheFile>>,
     inner: BTreeMap<CodeModuleId, CfiModule>,
 }
 
 impl CfiCacheModules {
     /// Creates a new CFI cache entries for code modules.
-    fn new<'a>(cfi_caches: impl Iterator<Item = &'a CfiCacheResult>) -> Self {
-        let inner = cfi_caches
-            .map(|(code_id, cache_result)| {
-                let cfi_module = match cache_result {
-                    Ok(cfi_cache) => {
-                        let cfi_status = match cfi_cache.status() {
-                            CacheStatus::Positive => ObjectFileStatus::Found,
-                            CacheStatus::Negative => ObjectFileStatus::Missing,
-                            CacheStatus::Malformed => {
-                                let err = CfiCacheError::ObjectParsing(ObjectError::Malformed);
-                                log::warn!("Error while parsing cficache: {}", LogError(&err));
-                                ObjectFileStatus::from(&err)
-                            }
-                        };
-                        let cfi_path = match cfi_cache.status() {
-                            CacheStatus::Positive => Some(cfi_cache.path().to_owned()),
-                            _ => None,
-                        };
-                        CfiModule {
-                            features: cfi_cache.features(),
-                            cfi_status,
-                            cfi_path,
-                            cfi_candidates: cfi_cache.candidates().clone(), // TODO(flub): fix clone
-                            ..Default::default()
+    fn new() -> Self {
+        Self {
+            cache_files: vec![],
+            inner: Default::default(),
+        }
+    }
+
+    /// Check if the Cache already contains the module with the given `id`.
+    fn has_module(&self, id: &CodeModuleId) -> bool {
+        self.inner.contains_key(id)
+    }
+
+    /// Extend the CacheModules with the fetched caches represented by
+    /// [`CfiCacheResult`].
+    fn extend(&mut self, cfi_caches: Vec<CfiCacheResult>) {
+        // make sure to hold onto a reference to the CfiCacheFile, to make sure
+        // it will not be evicted in the middle of using it
```

Comment:
Why does it suddenly become this cache's job to keep these files alive?

Reply:
Previously, they were moved into the stackwalking method, and that had an explicit drop at the end to keep the cache alive. This just moves this around a bit.